### DISCUSSION TITLE
Welcome to java-spring-boot Discussions!

### DISCUSSION THREAD

**Discussion by makklays** [makklays/java-spring-boot#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by makklays** [comment]
About microservices 
https://habr.com/ru/articles/682720/ 
https://habr.com/ru/articles/484130/
https://coderlessons.com/articles/devops-articles/sozdanie-mikroservisov-s-ispolzovaniem-spring-boot-i-docker-chast-2

**Comment by makklays** [comment]
About REST API and application 
https://spring.io/guides/gs/spring-boot/ 
https://smilyk1982.medium.com/restapi-java-spring-%D1%87%D0%B0%D1%81%D1%82%D1%8C-7-%D0%BF%D1%80%D0%BE%D0%B2%D0%B5%D1%80%D0%BA%D0%B0-api-%D0%BF%D1%80%D0%B8-%D0%BF%D0%BE%D0%BC%D0%BE%D1%89%D0%B8-postman-46f590f81a3e

And https://spring.io/

**Comment by makklays** [comment]
Realized spring-web 
my screens:
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/7515d683-e90d-4aa2-a0ad-e5661e3a2587)

![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/63626610-b7a1-4e97-94c2-3652d0dc3409)

![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/58cccb36-8984-43d3-aba1-49f9cb3b9eb3)

**Comment by makklays** [comment]
Have several URLs at REST API  and screens from results:
- http://127.0.0.1:8080/api/v1/currencies/USD

![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/5328881a-c08a-4bea-8fdb-6c238f7b7743)

- http://127.0.0.1:8080/api/v1/currencies/logs/2024-01-01/2024-01-31

**Comment by makklays** [comment]
Very interesting vacancy -  description:

Requisitos mínimos
Experiencia con: Java, Spring Boot, Oracle Weblogic, Jboss, Oracle, SonarQube, Jenkins

IMPRESCINDIBLE aportar una de estas certificaciones:
- Oracle Certified Professional, Java SE 8 Programmer
- Oracle Certified Professional, Java SE 11 Developer
- Oracle Certified Professional, Java SE 17 Developer
- Oracle Certified Professional, Java EE 7 Application Developer
- Oracle Database PL/SQL Developer Certified Profesional
- VMWare Spring Professional Develop
- Zend Certified Engineer
- Advanced Certified Scrum Developer
- Certified Scrum Professional for Developers
- Prince2 Practitioner Exam
- PMP

This is real?

**Comment by makklays** [comment]
Add code for get currency from National Bank and insert in my table `currencies` (for example)

Have next result:
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/be0a329a-75b5-4173-ac3c-832601b2fb9b)
F5
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/a7dbff16-fa66-411c-b017-1c77cb567de9)

Other currency, for example: ILS
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/dbc84e86-9f99-44b1-abdc-10d72a1ab631)

And list of currencies by period (date_start, date_end) from my table `currencies` (for example):
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/0723c0dc-c7dd-44e9-9911-75e9934f0654)

**Comment by makklays** [comment]
Beans.

EJB – Enterprise Java Beans.

Have 3 types:
- Entity Bean
- Session Bens (Stateless Session Bean and Statefull Session Bean)
- POJO (Plain Old Java Object) – старый обычный Java-объект

Objects:
- Entity 
- DAO (insert entity, update entity, etc)

**Comment by makklays** [comment]
Example of code (Repository use in Service -- Service use in Utils):

```
import org.springframework.stereotype.Service;
import ru.java.rush.repositories.FruitRepository;

@Service
public class FruitService {

    private final FruitRepository fruitRepository;  // final переменная репозитория

    public FruitService(FruitRepository fruitRepository) {  // внедрили зависимость через конструктор
          this.fruitRepository = fruitRepository;
    }
}
```

Repositories (JPA):
```
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import ru.java.rush.entities.FruitEntity;

@Repository//помечаем что этот бин - репозиторий
public interface FruitRepository extends JpaRepository<FruitEntity,Integer> {
     //репозиторий является интерфейсом, который наследуется от другого интерфейса JpaRepository<>
     //для него необходимо указать с какой сущность он должен работать, у нас это FruitEntity
     //и тип данных у поля id данной сущности, у нас это Integer
}
```

Example use repository in class:
```
class MyClient {

    private final PersonRepository repository;

    MyClient(PersonRepository repository) {
        this.repository = repository;
    }

    void doSomething() {
        List<Person> persons = repository.findByLastname("Sally");
    }
}
```

**Comment by upircloud** [comment]
Please have a look changes for hibernate also for controller... https://github.com/makklays/java-spring-boot/commit/4f0cb72458c07c9996b6dc7890518a7947907f73
Please reimplent other controller mappers and  services like i did it for inserting and getting currencies

**Reply by makklays** [reply]
Viewed and :
- added a new field 'created_at' for tables 'currencies';
- added new methods for Repository;
- changed sql-requests for database;

**Comment by makklays** [comment]
In git branch feature/rest-api-bank : 

![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/224eb418-98a2-4636-a6f7-31c4fc43ec41)

**Reply by upircloud** [reply]
Have you done ```git pull``` before committing?

**Reply by makklays** [reply]
Of course ) 
Deleted and did create new local branch 'feature/rest-api-bank' from remote branch 'origin feature/rest-api-bank'

**Comment by makklays** [comment]
After your commit on branch feature/rest-api-bank:

![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/8dbc9b89-d3f0-4938-a185-6db1cc5e867e)

**Comment by makklays** [comment]
stub - функции-заглушки
mock - объекта-имитация
unit - модульный тест

**Comment by makklays** [comment]
- DI на основе конструктора или на основе сеттера?
- Поскольку можно смешивать DI на основе конструктора и DI на основе сеттера, хорошим практическим правилом является использование конструкторов для обязательных зависимостей, а сеттеров или методов конфигурации - для необязательных зависимостей. Spring в целом поддерживает внедрение через конструктор, поскольку это позволяет реализовать компоненты приложения в виде неизменяемых объектов и гарантирует, что необходимые зависимости не будут null.

**Comment by makklays** [comment]
Внедрение зависимости через конструктор:
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/dddac246-60d0-4081-ae49-3892ffb04f4b)

Внедрение зависимости через сеттер:
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/ac4f087f-c4d2-4d56-b23b-622e687a365a)

-DI на основе сеттера (setter) выполняется путем вызова контейнером сеттеров для бинов после вызова конструктора без аргументов или статического фабричного метода без аргументов для создания экземпляра бина.

Внедрение зависимости через статический фабричный метод (через статический конструктор ))):
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/67bb8e0d-2fd0-4aa7-8599-e35b70a2d33b)

**Comment by makklays** [comment]
Add DataSource for H2: 
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/e19aea42-411d-4916-acb2-fc90674e3ae3)

Add DataSource for MySQL:
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/ec03598d-5c84-47f6-9205-a8b946955612)

Add DataSource for H2 (for tests): 
![изображение](https://github.com/makklays/java-spring-boot/assets/12418074/72b2efcc-24f1-4575-bfa5-0cc33cf28222)

EmbeddedDatabaseBuilder or config - what different between ?

**Reply by upircloud** [reply]
Actually, I have never used EmbeddedDatabaseBuilder, so I don't know what better, probably if it working for test, it's enough!

**Comment by upircloud** [comment]
So then just configure EmbeddedDatabaseBuilder for test purpose only, and configuration datasource via Hikari only for runtime.

**Comment by upircloud** [comment]
Exceptions https://www.baeldung.com/java-exceptions

**Comment by upircloud** [comment]
frontend java https://www.baeldung.com/jsf-servlet-jsp
